File: apps/dashboard/static/js/accounts.js
/* ==================================================================
 * ICYQuant Dashboard — Broker Account / Position Sync client (Commit 015)
 *
 * §16 / §17 — the front end talks to ICYQuant's own API only:
 *
 *     Dashboard → /api/accounts/* → AccountService → Snapshot
 *
 * It never reaches a broker.  That is why this file has no vendor
 * vocabulary in it: it knows the §12 status vocabulary, the §10
 * reconciliation verdicts and the response shapes, and nothing else.
 *
 * Two rules are encoded here rather than in each panel:
 *
 *   * **A verdict is not a colour.**  STATUS / HEALTH / RECONCILIATION
 *     tone mapping lives in one place, so a MISMATCH can never be painted
 *     green by a component that forgot the mapping.
 *   * **An absent number stays absent.**  `null` renders as "—", never as
 *     0, so an unvalued position (§18) is visibly unvalued instead of
 *     looking like a zero-value holding.
 */
(function () {
  "use strict";

  var API = window.ICY_API;

  /* ── §12 account status vocabulary ─────────────────────────────── */

  var STATUS_META = {
    CONNECTED: { tone: "info", label: "CONNECTED", zh: "已连接" },
    SYNCING: { tone: "warning", label: "SYNCING", zh: "同步中" },
    SYNCED: { tone: "profit", label: "SYNCED", zh: "已同步" },
    STALE: { tone: "warning", label: "STALE", zh: "数据过期" },
    MISMATCH: { tone: "danger", label: "MISMATCH", zh: "持仓不符" },
    ERROR: { tone: "danger", label: "ERROR", zh: "错误" },
    OFFLINE: { tone: "neutral", label: "OFFLINE", zh: "离线" },
  };

  /* §14 roll-up health (Monitor vocabulary, Commit 013). */
  var HEALTH_META = {
    HEALTHY: { tone: "profit", label: "HEALTHY", zh: "正常" },
    DEGRADED: { tone: "warning", label: "DEGRADED", zh: "降级" },
    BLOCKED: { tone: "danger", label: "BLOCKED", zh: "已阻断" },
    OFFLINE: { tone: "neutral", label: "OFFLINE", zh: "离线" },
  };

  /* §10 reconciliation verdicts. */
  var RECON_META = {
    RECONCILED: { tone: "profit", label: "PASS", zh: "一致" },
    MISMATCH: { tone: "danger", label: "MISMATCH", zh: "不一致" },
    MISSING_BROKER: { tone: "danger", label: "MISSING @BROKER", zh: "券商缺失" },
    MISSING_LEDGER: { tone: "danger", label: "MISSING @LEDGER", zh: "账本缺失" },
    INVALID: { tone: "danger", label: "INVALID", zh: "无效" },
  };

  var CHECK_LABELS = {
    ACCOUNT_BALANCE: "账户资金",
    POSITION_QUANTITY: "持仓数量",
    AVAILABLE_QUANTITY: "可用数量",
    AVERAGE_COST: "持仓成本",
    MARKET_VALUE: "持仓市值",
  };

  function _upper(v) {
    return String(v == null ? "" : v).toUpperCase();
  }

  function meta(table, key, fallbackTone) {
    var m = table[_upper(key)];
    return m || { tone: fallbackTone || "neutral", label: _upper(key) || "UNKNOWN", zh: "" };
  }

  function statusMeta(status) { return meta(STATUS_META, status); }
  function healthMeta(health) { return meta(HEALTH_META, health); }
  function reconMeta(status) { return meta(RECON_META, status, "warning"); }
  function checkLabel(check) { return CHECK_LABELS[_upper(check)] || _upper(check); }

  /* §11 — blocking is a consequence of the verdict, derived here once. */
  function isBlocked(syncStatus) {
    if (!syncStatus) return false;
    if (syncStatus.new_orders_blocked === true) return true;
    return _upper(syncStatus.status) === "MISMATCH";
  }

  /* ── number / time formatting ──────────────────────────────────── */

  var DASH = "—";

  function _num(v) {
    if (v === null || v === undefined || v === "") return null;
    var n = Number(v);
    return isFinite(n) ? n : null;
  }

  function fmtMoney(v, decimals) {
    var n = _num(v);
    if (n === null) return DASH;
    return "¥" + n.toLocaleString("zh-CN", {
      minimumFractionDigits: decimals == null ? 2 : decimals,
      maximumFractionDigits: decimals == null ? 2 : decimals,
    });
  }

  function fmtSignedMoney(v) {
    var n = _num(v);
    if (n === null) return DASH;
    return (n >= 0 ? "+¥" : "-¥") + Math.abs(n).toLocaleString("zh-CN", {
      minimumFractionDigits: 2,
      maximumFractionDigits: 2,
    });
  }

  function fmtPrice(v, decimals) {
    var n = _num(v);
    if (n === null) return DASH;
    return n.toFixed(decimals == null ? 3 : decimals);
  }

  function fmtQty(v) {
    var n = _num(v);
    if (n === null) return DASH;
    return n.toLocaleString("zh-CN", { maximumFractionDigits: 0 });
  }

  function fmtPct(v) {
    var n = _num(v);
    if (n === null) return DASH;
    // The API sends a ratio (0.0234 → "+2.34%").
    var scaled = Math.abs(n) <= 1.5 ? n * 100 : n;
    return (scaled >= 0 ? "+" : "") + scaled.toFixed(2) + "%";
  }

  function fmtLatency(ms) {
    var n = _num(ms);
    if (n === null) return DASH;
    return n >= 1000 ? (n / 1000).toFixed(2) + "s" : Math.round(n) + "ms";
  }

  function fmtAge(seconds) {
    var n = _num(seconds);
    if (n === null) return DASH;
    if (n < 60) return Math.round(n) + "s";
    if (n < 3600) return Math.floor(n / 60) + "m " + Math.round(n % 60) + "s";
    return Math.floor(n / 3600) + "h " + Math.floor((n % 3600) / 60) + "m";
  }

  /* §13 — always show the broker's clock and ours, never just ours. */
  function fmtTime(iso) {
    if (!iso) return DASH;
    var d = new Date(iso);
    if (isNaN(d.getTime())) return String(iso);
    return d.toLocaleTimeString("zh-CN", { hour12: false }) +
      "." + String(d.getMilliseconds()).padStart(3, "0");
  }

  function fmtDateTime(iso) {
    if (!iso) return DASH;
    var d = new Date(iso);
    if (isNaN(d.getTime())) return String(iso);
    return d.toLocaleString("zh-CN", { hour12: false });
  }

  /* ── §15 API client ────────────────────────────────────────────── */

  var accountsApi = {
    base: "/accounts",

    list: function () {
      return API.get("/accounts");
    },

    detail: function (accountId) {
      return API.get("/accounts/" + encodeURIComponent(accountId));
    },

    getBalance: function (accountId) {
      return API.get("/accounts/" + encodeURIComponent(accountId) + "/balance");
    },

    getPositions: function (accountId) {
      return API.get("/accounts/" + encodeURIComponent(accountId) + "/positions");
    },

    getSyncStatus: function (accountId) {
      return API.get("/accounts/" + encodeURIComponent(accountId) + "/sync-status");
    },

    getSnapshots: function (accountId, limit) {
      return API.get("/accounts/" + encodeURIComponent(accountId) + "/snapshots", {
        limit: limit || 20,
      });
    },

    health: function () {
      return API.get("/accounts/health");
    },

    connect: function () {
      return API.post("/accounts/connect");
    },

    disconnect: function () {
      return API.post("/accounts/disconnect");
    },

    /* Operator action (§6) — the Dashboard never triggers this on a poll. */
    sync: function (accountId, connect) {
      return API.post(
        "/accounts/" + encodeURIComponent(accountId) + "/sync" +
          (connect === false ? "?connect=false" : "")
      );
    },
  };

  /* ── §16 table schemas ─────────────────────────────────────────── */

  var POSITION_COLUMNS = [
    { key: "symbol", label: "Symbol", sortable: true },
    { key: "name", label: "Name / 名称" },
    { key: "quantity", label: "Qty / 数量", numeric: true, format: fmtQty },
    { key: "available_quantity", label: "Available / 可用", numeric: true, format: fmtQty },
    { key: "frozen_quantity", label: "Frozen / 冻结", numeric: true, format: fmtQty },
    { key: "average_cost", label: "Cost / 成本", numeric: true, format: fmtPrice },
    { key: "market_price", label: "Price / 现价", numeric: true, format: fmtPrice },
    { key: "market_value", label: "Market Value / 市值", numeric: true, format: fmtMoney },
    {
      key: "unrealized_pnl",
      label: "P&L / 盈亏",
      numeric: true,
      format: fmtSignedMoney,
      color: function (v) { var n = _num(v); return n == null ? "" : (n >= 0 ? "pos" : "neg"); },
    },
    {
      key: "unrealized_pnl_pct",
      label: "P&L % / 收益率",
      numeric: true,
      format: fmtPct,
      color: function (v) { var n = _num(v); return n == null ? "" : (n >= 0 ? "pos" : "neg"); },
    },
    {
      key: "status",
      label: "Status",
      format: function (v) {
        // §3 / §21 — an INVALID position must be visible, not silently listed.
        var m = _upper(v) === "INVALID"
          ? { tone: "danger", label: "INVALID" }
          : { tone: "profit", label: "VALID" };
        return UI.statusPill(m.label, m.tone);
      },
    },
  ];

  var SNAPSHOT_COLUMNS = [
    { key: "sequence", label: "#", numeric: true, format: function (v) { return v == null ? DASH : "#" + v; } },
    { key: "snapshot_id", label: "Snapshot ID" },
    { key: "timestamp", label: "Broker Time / 券商时间", format: fmtTime },
    { key: "received_timestamp", label: "Received / 接收时间", format: fmtTime },
    { key: "sync_latency_ms", label: "Latency", numeric: true, format: fmtLatency },
    { key: "position_count", label: "Positions", numeric: true, format: fmtQty },
    { key: "source", label: "Source" },
    {
      key: "reconciliation_status",
      label: "Reconciliation",
      format: function (v) {
        if (!v) return DASH;
        var m = reconMeta(v);
        return UI.statusPill(m.label, m.tone);
      },
    },
  ];

  var RECON_ITEM_COLUMNS = [
    { key: "symbol", label: "Symbol" },
    { key: "check", label: "Check / 检查项", format: checkLabel },
    { key: "broker_value", label: "Broker / 券商", numeric: true, format: function (v, row) { return _fmtByCheck(row.check, v); } },
    { key: "ledger_value", label: "Ledger / 账本", numeric: true, format: function (v, row) { return _fmtByCheck(row.check, v); } },
    { key: "difference", label: "Diff / 差异", numeric: true, format: function (v, row) { return _fmtByCheck(row.check, v); },
      color: function (v) { var n = _num(v); return n == null || n === 0 ? "" : "neg"; } },
    {
      key: "status",
      label: "Status",
      format: function (v) {
        var m = reconMeta(v);
        return UI.statusPill(m.label, m.tone);
      },
    },
  ];

  function _fmtByCheck(check, value) {
    if (_upper(check) === "AVERAGE_COST") return fmtPrice(value);
    if (_upper(check) === "POSITION_QUANTITY" || _upper(check) === "AVAILABLE_QUANTITY") {
      return fmtQty(value);
    }
    return fmtMoney(value);
  }

  window.ICY_ACCOUNTS = {
    api: accountsApi,
    STATUS: STATUS_META,
    HEALTH: HEALTH_META,
    RECONCILIATION: RECON_META,
    statusMeta: statusMeta,
    healthMeta: healthMeta,
    reconMeta: reconMeta,
    checkLabel: checkLabel,
    isBlocked: isBlocked,
    POSITION_COLUMNS: POSITION_COLUMNS,
    SNAPSHOT_COLUMNS: SNAPSHOT_COLUMNS,
    RECON_ITEM_COLUMNS: RECON_ITEM_COLUMNS,
    fmtMoney: fmtMoney,
    fmtSignedMoney: fmtSignedMoney,
    fmtPrice: fmtPrice,
    fmtQty: fmtQty,
    fmtPct: fmtPct,
    fmtLatency: fmtLatency,
    fmtAge: fmtAge,
    fmtTime: fmtTime,
    fmtDateTime: fmtDateTime,
    DASH: DASH,
  };

  /* Alias used by the spec's components. */
  window.accountsApi = accountsApi;
})();

File: apps/api/routers/accounts.py
"""Account / Position API — the single standard entry point (Commit 015 §15).

Mounted at ``/api/accounts``::

    GET  /                                 Account registry + connection
    GET  /health                           §14 sync health roll-up
    GET  /{account_id}                     Account + status + balance + positions
    GET  /{account_id}/balance             Balance (§2)
    GET  /{account_id}/positions           Positions + reconciliation (§3 / §10)
    GET  /{account_id}/sync-status         §12 status, §13 latency, §7 snapshots
    GET  /{account_id}/snapshots           Snapshot history (§7)
    POST /connect | /disconnect            Channel lifecycle (OPERATOR/ADMIN)
    POST /{account_id}/sync                Pull one snapshot (OPERATOR/ADMIN)

Design rules honoured here:

* **The API never talks to the broker** (§15).  Routes do no data work:
  they validate HTTP input, call
  :data:`~services.account.sync.service.account_service` and serialise a
  Pydantic model.  §17 holds by construction — Dashboard → API →
  AccountService → Snapshot, so the front end is never coupled to a
  vendor.
* **Reads are side-effect free.**  A ``GET`` serves the last retained
  snapshot; it never opens a broker connection.  Pulling fresh state is
  an explicit operator action (``POST /{account_id}/sync``), which is why
  ``/health`` is safe to poll.
* **Quality is never hidden.**  Every position payload carries ``status``
  (VALID/INVALID) and every account payload carries the §10
  reconciliation verdict, so a MISMATCH account is served *as* MISMATCH
  (§11) rather than as a plausible-looking number.
* **No snapshot yet is honest.**  404 ``ACCOUNT_SNAPSHOT_NOT_FOUND``
  rather than a fabricated empty account.
"""
from __future__ import annotations

import logging
from typing import Optional

from fastapi import APIRouter, Depends, Query, Response
from fastapi.responses import JSONResponse

from apps.api.schemas.accounts import (
    AccountDetailResponse,
    AccountHealthResponse,
    AccountListResponse,
    AccountSchema,
    AccountsErrorResponse,
    BalanceSchema,
    ConnectionActionResponse,
    PositionListResponse,
    PositionSchema,
    ReconciliationItemSchema,
    ReconciliationSchema,
    SnapshotListResponse,
    SnapshotSchema,
    SyncActionResponse,
    SyncStatusResponse,
)
from apps.dashboard.auth import Principal, require_roles
from services.account.domain.account import Account
from services.account.domain.balance import AccountBalance
from services.account.domain.exceptions import AccountError
from services.account.domain.position import Position
from services.account.domain.snapshot import AccountSnapshot
from services.account.sync.exceptions import AccountSyncError
from services.account.sync.service import (
    AccountSyncService,
    account_service,
)
from services.market_data.universe import universe

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/accounts", tags=["accounts"])

_MAX_SNAPSHOT_LIMIT = 500

_ERROR_RESPONSES: dict = {
    400: {
        "model": AccountsErrorResponse,
        "description": "Bad request / unknown provider",
    },
    404: {
        "model": AccountsErrorResponse,
        "description": "Unknown account or no snapshot",
    },
    409: {
        "model": AccountsErrorResponse,
        "description": "Account / broker state conflict",
    },
    422: {
        "model": AccountsErrorResponse,
        "description": "Broker payload invalid",
    },
    502: {
        "model": AccountsErrorResponse,
        "description": "Broker account request failed",
    },
    503: {
        "model": AccountsErrorResponse,
        "description": "Broker channel unavailable",
    },
}


def accounts_error_handler(request, exc: AccountError):
    """Render any account error as one envelope (§15).

    :class:`~services.account.sync.exceptions.AccountSyncError` carries its
    own ``status_code``; a pure domain error falls back to 400 — the
    request was well-formed but the account data does not add up.
    """
    status_code = getattr(exc, "status_code", 400)
    return JSONResponse(status_code=status_code, content=exc.as_dict())


def _service() -> AccountSyncService:
    """Single shared service instance (§15 / §17)."""
    return account_service


# ══════════════════════════════════════════════════════════════════
# Serialisation helpers — domain object → Pydantic model
# ══════════════════════════════════════════════════════════════════


def _account_schema(account: Account) -> AccountSchema:
    return AccountSchema(
        account_id=account.account_id,
        name=account.name,
        broker=account.broker,
        currency=account.currency,
        status=account.status,
    )


def _balance_schema(balance: Optional[AccountBalance]) -> Optional[BalanceSchema]:
    if balance is None:
        return None
    return BalanceSchema(
        account_id=balance.account_id,
        currency=balance.currency,
        cash=float(balance.cash),
        available_cash=float(balance.available_cash),
        frozen_cash=float(balance.frozen_cash),
        market_value=float(balance.market_value),
        total_asset=float(balance.total_asset),
        buying_power=float(balance.buying_power),
        broker_timestamp=balance.broker_timestamp,
        received_timestamp=balance.received_timestamp,
        consistent=balance.consistent,
    )


def _position_schema(position: Position) -> PositionSchema:
    """One position (§3) — plus the Instrument Master name for §16's table.

    ``name`` is a display join on the §2 registry; the account model itself
    stays symbol-only, and §20's rule holds: the universe is not the
    position list, so a name may legitimately be ``None``.
    """
    instrument = universe.get(position.symbol)
    return PositionSchema(
        symbol=position.symbol,
        exchange=position.exchange,
        name=instrument.name if instrument else None,
        quantity=float(position.quantity),
        available_quantity=float(position.available_quantity),
        frozen_quantity=float(position.frozen_quantity),
        average_cost=float(position.average_cost),
        market_price=(
            None if position.market_price is None else float(position.market_price)
        ),
        market_value=(
            None if position.market_value is None else float(position.market_value)
        ),
        unrealized_pnl=(
            None if position.unrealized_pnl is None else float(position.unrealized_pnl)
        ),
        unrealized_pnl_pct=(
            None
            if position.unrealized_pnl_pct is None
            else float(position.unrealized_pnl_pct)
        ),
        status=position.status,
        consistent=position.consistent,
        broker_timestamp=position.broker_timestamp,
        received_timestamp=position.received_timestamp,
    )


def _reconciliation_schema(payload: Optional[dict]) -> Optional[ReconciliationSchema]:
    if not payload:
        return None
    return ReconciliationSchema(
        account_id=payload["account_id"],
        snapshot_id=payload["snapshot_id"],
        timestamp=payload["timestamp"],
        status=payload["status"],
        outcome=payload["outcome"],
        checked=payload["checked"],
        matched=payload["matched"],
        mismatch=payload["mismatch"],
        missing_broker=payload["missing_broker"],
        missing_ledger=payload["missing_ledger"],
        invalid=payload["invalid"],
        mismatch_count=payload["mismatch_count"],
        items=[
            ReconciliationItemSchema(**item) for item in payload.get("items", [])
        ],
    )


def _sync_status_schema(
    service: AccountSyncService, status: dict
) -> SyncStatusResponse:
    config = service.config
    account_id = status.get("account_id") or ""
    health = (
        service.health(account_id).get("health") if account_id else "OFFLINE"
    )
    return SyncStatusResponse(
        account_id=account_id,
        status=status.get("status", "OFFLINE"),
        health=health or "OFFLINE",
        connection_state=status.get("connection_state", "DISCONNECTED"),
        broker_connected=bool(status.get("broker_connected")),
        provider=status.get("provider", service.adapter.provider_name),
        source=status.get("source", service.source),
        last_successful_sync=status.get("last_successful_sync"),
        last_sync_age_seconds=status.get("last_sync_age_seconds"),
        sync_latency_ms=status.get("sync_latency_ms"),
        stale_after_seconds=config.max_sync_age_seconds,
        sync_interval_seconds=config.sync_interval_seconds,
        position_count=int(status.get("position_count") or 0),
        invalid_count=int(status.get("invalid_count") or 0),
        mismatch_count=int(status.get("mismatch_count") or 0),
        snapshot_count=int(status.get("snapshot_count") or 0),
        reconciliation_status=status.get("reconciliation_status"),
        reconciliation_outcome=status.get("reconciliation_outcome"),
        new_orders_blocked=bool(status.get("new_orders_blocked")),
    )


def _snapshot_schema(snapshot: AccountSnapshot) -> SnapshotSchema:
    reconciliation = snapshot.reconciliation or {}
    return SnapshotSchema(
        snapshot_id=snapshot.snapshot_id,
        account_id=snapshot.account_id,
        sequence=snapshot.sequence,
        status=snapshot.status,
        source=snapshot.source,
        timestamp=snapshot.timestamp,
        received_timestamp=snapshot.received_timestamp,
        sync_latency_ms=snapshot.sync_latency_ms,
        position_count=snapshot.position_count,
        reconciliation_status=reconciliation.get("status"),
        reconciliation_outcome=reconciliation.get("outcome"),
        balance=_balance_schema(snapshot.balance),
    )


def _sync_action_schema(snapshot: AccountSnapshot) -> SyncActionResponse:
    return SyncActionResponse(
        account_id=snapshot.account_id,
        status=snapshot.status,
        snapshot_id=snapshot.snapshot_id,
        sequence=snapshot.sequence,
        position_count=snapshot.position_count,
        sync_latency_ms=snapshot.sync_latency_ms,
        reconciliation=_reconciliation_schema(snapshot.reconciliation),
    )


# ══════════════════════════════════════════════════════════════════
# ① Account registry
# ══════════════════════════════════════════════════════════════════


@router.get(
    "",
    response_model=AccountListResponse,
    responses=_ERROR_RESPONSES,
    summary="Registered broker accounts",
)
def list_accounts(
    principal: Principal = Depends(require_roles()),
):
    """Every registered account with the §5 connection state (§15)."""
    service = _service()
    accounts = service.accounts()
    return AccountListResponse(
        accounts=[_account_schema(a) for a in accounts],
        count=len(accounts),
        provider=service.adapter.provider_name,
        connection_state=service.connection_state(),
        broker_connected=service.is_connected(),
        source=service.source,
        new_orders_blocked=service.new_orders_blocked(),
    )


@router.get(
    "/health",
    response_model=AccountHealthResponse,
    summary="Account / position sync health",
)
def get_account_health(
    principal: Principal = Depends(require_roles()),
):
    """§14 roll-up for Commit 013 — non-mutating, safe to poll.

    Declared before ``/{account_id}`` so ``health`` is never read as an
    account id.
    """
    return _service().health_report()


# ══════════════════════════════════════════════════════════════════
# ② Lifecycle (operator actions)
# ══════════════════════════════════════════════════════════════════


@router.post(
    "/connect",
    response_model=ConnectionActionResponse,
    responses=_ERROR_RESPONSES,
    summary="Open the broker account channel",
)
def connect_accounts(
    principal: Principal = Depends(require_roles("OPERATOR", "ADMIN")),
):
    """Open the account channel (§5).  Idempotent at the adapter level."""
    service = _service()
    service.connect()
    return ConnectionActionResponse(
        provider=service.adapter.provider_name,
        connection_state=service.connection_state(),
        broker_connected=service.is_connected(),
        accounts=[a.account_id for a in service.accounts()],
    )


@router.post(
    "/disconnect",
    response_model=ConnectionActionResponse,
    responses=_ERROR_RESPONSES,
    summary="Close the broker account channel",
)
def disconnect_accounts(
    principal: Principal = Depends(require_roles("OPERATOR", "ADMIN")),
):
    """Close the account channel — snapshots already stored stay readable."""
    service = _service()
    service.disconnect()
    return ConnectionActionResponse(
        provider=service.adapter.provider_name,
        connection_state=service.connection_state(),
        broker_connected=service.is_connected(),
        accounts=[a.account_id for a in service.accounts()],
    )


# ══════════════════════════════════════════════════════════════════
# ③ Account detail
# ══════════════════════════════════════════════════════════════════


@router.get(
    "/{account_id}",
    response_model=AccountDetailResponse,
    responses=_ERROR_RESPONSES,
    summary="One account with its latest state",
)
def get_account(
    account_id: str,
    principal: Principal = Depends(require_roles()),
):
    """Account + sync status + latest balance / positions (§15).

    A registered account that has never synced answers with ``OFFLINE``
    status and empty positions — honest, not fabricated.
    """
    service = _service()
    account = service.account(account_id)
    latest = service.latest(account_id)
    return AccountDetailResponse(
        account=_account_schema(account),
        sync_status=_sync_status_schema(service, service.status(account_id)),
        balance=_balance_schema(latest.balance if latest else None),
        positions=[_position_schema(p) for p in latest.positions] if latest else [],
        reconciliation=_reconciliation_schema(service.reconciliation(account_id)),
    )


@router.get(
    "/{account_id}/balance",
    response_model=BalanceSchema,
    responses=_ERROR_RESPONSES,
    summary="Latest account balance",
)
def get_balance(
    account_id: str,
    principal: Principal = Depends(require_roles()),
):
    """Cash / available / frozen / market value / total asset (§2)."""
    service = _service()
    service.account(account_id)
    return _balance_schema(service.balance(account_id))


@router.get(
    "/{account_id}/positions",
    response_model=PositionListResponse,
    responses=_ERROR_RESPONSES,
    summary="Latest positions",
)
def get_positions(
    account_id: str,
    principal: Principal = Depends(require_roles()),
):
    """Positions with the §10 reconciliation verdict attached (§15).

    The response carries ``status`` and ``reconciliation`` next to the
    numbers, so a caller physically cannot read a MISMATCH account as if
    it were reconciled.
    """
    service = _service()
    service.account(account_id)
    snapshot = service.require_snapshot(account_id)
    positions = list(snapshot.positions)
    return PositionListResponse(
        account_id=account_id,
        status=snapshot.status,
        snapshot_id=snapshot.snapshot_id,
        snapshot_timestamp=snapshot.timestamp,
        count=len(positions),
        positions=[_position_schema(p) for p in positions],
        positions_market_value=snapshot.total_market_value,
        total_cost=float(sum(p.average_cost * p.quantity for p in positions)),
        total_unrealized_pnl=float(
            sum(
                p.unrealized_pnl
                for p in positions
                if p.unrealized_pnl is not None
            )
        ),
        reconciliation=_reconciliation_schema(service.reconciliation(account_id)),
    )


@router.get(
    "/{account_id}/sync-status",
    response_model=SyncStatusResponse,
    responses=_ERROR_RESPONSES,
    summary="Sync status, latency and staleness",
)
def get_sync_status(
    account_id: str,
    principal: Principal = Depends(require_roles()),
):
    """§12 status, §13 latency, §7 snapshot count and §11 blocking flag."""
    service = _service()
    service.account(account_id)
    return _sync_status_schema(service, service.status(account_id))


@router.get(
    "/{account_id}/snapshots",
    response_model=SnapshotListResponse,
    responses=_ERROR_RESPONSES,
    summary="Retained snapshot history",
)
def get_snapshots(
    account_id: str,
    limit: int = Query(
        50,
        ge=1,
        le=_MAX_SNAPSHOT_LIMIT,
        description="How many snapshots to return, newest first (§7)",
    ),
    principal: Principal = Depends(require_roles()),
):
    """Newest-first snapshot chain — snapshots are never overwritten (§7).

    Answering "why did the system think I held 10,000 shares?" means
    re-reading the snapshot from that moment, so history is never
    compacted here.
    """
    service = _service()
    service.account(account_id)
    snapshots = service.snapshots(account_id, limit=limit)
    return SnapshotListResponse(
        account_id=account_id,
        count=len(snapshots),
        limit=limit,
        snapshots=[_snapshot_schema(s) for s in snapshots],
    )


# ══════════════════════════════════════════════════════════════════
# ④ Manual snapshot pull (operator action, §6)
# ══════════════════════════════════════════════════════════════════


@router.post(
    "/{account_id}/sync",
    response_model=SyncActionResponse,
    responses=_ERROR_RESPONSES,
    summary="Pull one account + position snapshot",
)
def sync_account(
    account_id: str,
    response: Response,
    connect: bool = Query(
        True,
        description="Open the channel first when it is not already connected",
    ),
    principal: Principal = Depends(require_roles("OPERATOR", "ADMIN")),
):
    """Take a snapshot now (§6).

    Snapshot sync is a scheduled job in production; this endpoint is the
    operator's manual equivalent, so the Dashboard and Commit 017's E2E
    validation can drive the real pipeline instead of a fixture.  It goes
    through the service — the route still never calls a vendor SDK
    (§15 / §17).
    """
    service = _service()
    service.account(account_id)
    if connect and not service.is_connected():
        service.connect()
    snapshot = service.sync(account_id)
    response.status_code = 202
    return _sync_action_schema(snapshot)


__all__ = ["router", "accounts_error_handler"]
